### DISCUSSION TITLE
Issues with Reconnecting to Deepgram WebSocket After Closure and Reinitialization

### DISCUSSION THREAD

**Discussion by shrey231** [deepgram/community#525] [General help] [answered]
Hello,

I'm encountering a problem with reconnecting to a Deepgram WebSocket in a React component. Initially, upon component load, I can successfully establish a connection to Deepgram's WebSocket and transmit audio data. However, issues arise when I attempt to reconnect after having closed the connection.

Here's the workflow that leads to the problem:
1. On component initialization, the WebSocket connection is established successfully, and audio data is transmitted correctly to Deepgram's server and I receive a transcript from Deepgram.
2. When pausing the live transcription, I close the WebSocket and set its reference to null.
3. Upon resuming, I invoke the same function used for the initial connection to establish a new WebSocket connection.
4. After reinitialization, the WebSocket connection is established, but no audio data is transmitted to Deepgram's server. I have confirmed that the audio data being sent is valid and correctly formatted.

The main issue seems to be centered around the `handlePause` function, where pausing and then resuming does not resume data transmission to Deepgram, even though the WebSocket connection itself is reestablished without errors.

Below are the relevant code snippets showcasing how the WebSocket is initialized, closed, and then reinitialized:

   export const createDeepgramSocket = () => {
        const deepgramUrl = new URL("wss://api.deepgram.com/v1/listen");
      
        deepgramUrl.searchParams.append("punctuate", "true");
        deepgramUrl.searchParams.append("model", "nova-2"); 
      
        return new WebSocket(
          deepgramUrl.toString(),
          ["token", process.env.REACT_APP_MINDGRASP_V2_DEEPGRAM_KEY]
        );
      };

  const initializeWebSocket = async () => {
      if (!deepgramSocket || !deepgramSocket.current || deepgramSocket.current.readyState !== WebSocket.OPEN) {
  
        deepgramSocket.current = await createDeepgramSocket();
        let lastTimestamp = 0;
        
      
        deepgramSocket.current.onopen = async function () {
          console.log("WebSocket is open now.");
          if(!mediaRecorder.current) {
            const stream = await navigator.mediaDevices.getUserMedia({ audio: true });
              mediaRecorder.current = new MediaRecorder(stream);
              mediaRecorder.current.start(1000);
          }
          
          mediaRecorder.current.ondataavailable = function (e) {
            
            if (deepgramSocket && deepgramSocket.current && deepgramSocket.current.readyState === WebSocket.OPEN) {
              
              deepgramSocket.current.send(e.data);
            }
            audioChunks.current.push(e.data);
          };
  
        };
    
        deepgramSocket.current.onmessage = function (event) {
          console.log(event.data)
          if (!isPaused) {
            const response = JSON.parse(event.data);
            if (
              response.channel &&
              response.channel.alternatives &&
              response.channel.alternatives.length > 0
            ) {
              const transcript = response.channel.alternatives[0].transcript;
              if (totalSecondsRef.current - lastTimestamp >= 2) {
                let timestamp = formatTimestamp(totalSecondsRef.current);
         
                setRecordedText((prevText) => {
                  if (prevText.length === 0) {
                    return `[${timestamp}] ${transcript}`;
                  } else {
                    return `${prevText}\n[${timestamp}] ${transcript}`;
                  }
                });
                lastTimestamp = totalSecondsRef.current;
              } else {
                setRecordedText((prevText) => prevText + transcript + " ");
              }
     
            }
          }
        };
    
        deepgramSocket.current.onerror = (event) => {
          console.error("WebSocket error observed:", event);
        };
    
        deepgramSocket.current.onclose = (event) => {
          console.log("WebSocket is closed now. Reason:", event.reason);
          deepgramSocket.current = null;
        };
      }
    };
  

  async function startRecording() {
      try {
        start();
        setStartRecord(true);
        setIsPaused(false);
        initializeWebSocket()
        try {
          captureEvent("live_recording_started", { user_id: user.uid });
        } catch (err) {
          console.log("Failed to capture Live Recording...");
        }
  
        const remainingTimeInSeconds = user.add_ons.live_recording.credit_balance * 3600;
        
        const checkInterval = setInterval(() => {
          if (totalSecondsRef.current >= remainingTimeInSeconds) {
              stopRecording(true); 
              clearInterval(checkInterval); 
        
        
          }
        }, 1000);
        setIntervalId(checkInterval);
  
  
      } catch (error) {
        console.error("Error accessing the microphone", error);
      }
    }
  

   const handlePause = async () => {
      try {
  
        if (mediaRecorder.current) {
          if (isPaused) {
            // Resume recording if paused
            if (mediaRecorder.current.state === 'paused') {
              console.log("Resuming recording...");
              mediaRecorder.current.resume();
              console.log(`MediaRecorder state after resume: ${mediaRecorder.current.state}`);
        
              // Initialize WebSocket
              await initializeWebSocket();
              start()
            }
          } else {
            // Pause recording if active
            if (mediaRecorder.current.state === 'recording') {
    
              if (deepgramSocket.current && deepgramSocket.current.readyState === WebSocket.OPEN) {
  
                deepgramSocket.current.close(1000, 'Manual closure');
                deepgramSocket.current = null;
              }
              mediaRecorder.current.pause();
              console.log(`MediaRecorder state after pause: ${mediaRecorder.current.state}`);
              pause();
            }
          }
          setIsPaused(!isPaused);
        } 
      } catch (error) {
        console.error("Could not toggle pause state", error);
      }
    };

I would greatly appreciate any insights or suggestions on how to resolve this issue, ensuring that the WebSocket can transmit audio data to Deepgram's server upon reconnection.

Thank you in advance for your assistance!

**Comment by team-deepgram** [comment]
Thanks for asking your question about Deepgram! If you didn't already include it in your post, please be sure to add as much detail as possible so we can assist you efficiently, such as:
- The `request_id` if you have a question about your requests or transcription responses.
- The features you used or the full `api.deepgram.com` URL you sent your request to, including parameters.
- Any code snippets you can share.

**Comment by BrianHillis-DG** [comment, marked as answer]
Hey @shrey231

Try starting and stopping the `MediaRecorder` instead of pausing and resuming if you are going to close the socket when you want to trigger a "pause".

One small difference between `resume()` and `start()` in a `MediaRecorder` is `start()` applies header/metadata information to the first Blob.
